x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/wrapper/DlsFlsFeatureTrackingIndicesAccessControlWrapper.java

```diff
@@ -0,0 +1,72 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authz.accesscontrol.wrapper;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.authz.accesscontrol.IndicesAccessControl;
+
+import static org.elasticsearch.xpack.core.security.SecurityField.DOCUMENT_LEVEL_SECURITY_FEATURE;
+import static org.elasticsearch.xpack.core.security.SecurityField.FIELD_LEVEL_SECURITY_FEATURE;
+
+/**
+ * The wrapper of {@link IndicesAccessControl} which adds ability to track actual Document and Field Level Security feature usage.
+ */
+public class DlsFlsFeatureTrackingIndicesAccessControlWrapper {
+
+    private final XPackLicenseState licenseState;
+    private final boolean isDlsFlsEnabled;
+
+    public DlsFlsFeatureTrackingIndicesAccessControlWrapper(Settings settings, XPackLicenseState licenseState) {
+        this.licenseState = licenseState;
+        this.isDlsFlsEnabled = XPackSettings.DLS_FLS_ENABLED.get(settings);
+    }
+
+    public IndicesAccessControl wrap(IndicesAccessControl indicesAccessControl) {
+        if (isDlsFlsEnabled == false
+            || indicesAccessControl == null
+            || indicesAccessControl == IndicesAccessControl.ALLOW_NO_INDICES
+            || indicesAccessControl == IndicesAccessControl.DENIED
+            || indicesAccessControl == IndicesAccessControl.allowAll()
+            || indicesAccessControl instanceof DlsFlsFeatureUsageTracker) {
+            // Wrap only if it's not one of the statically defined nor already wrapped.
+            return indicesAccessControl;
+        } else {
+            return new DlsFlsFeatureUsageTracker(indicesAccessControl, licenseState);
+        }
+    }
+
+    private static class DlsFlsFeatureUsageTracker extends IndicesAccessControl {
+
+        private final XPackLicenseState licenseState;
+
+        private DlsFlsFeatureUsageTracker(IndicesAccessControl wrapped, XPackLicenseState licenseState) {
+            super(wrapped);
+            this.licenseState = licenseState;
+        }
+
+        @Override
+        public IndexAccessControl getIndexPermissions(String index) {
+            final IndexAccessControl iac = super.getIndexPermissions(index);
+            if (iac != null) {
+                trackUsage(iac);
+            }
+            return iac;
+        }
+
+        private void trackUsage(IndexAccessControl iac) {
+            if (iac.getDocumentPermissions().hasDocumentLevelPermissions()) {
+                DOCUMENT_LEVEL_SECURITY_FEATURE.check(licenseState);
+            }
+            if (iac.getFieldPermissions().hasFieldLevelSecurity()) {
+                FIELD_LEVEL_SECURITY_FEATURE.check(licenseState);
+            }
```

Comment:
What do you think if we assert that the features are licensed and that we log WARN if it isn't?

Reply:
Adding `assert` sounds like a good idea. I'm not sure if logging a warning makes sense here, since `DlsFlsLicenseRequestInterceptor` should fail the request before even reaching this code.

In theory, if the license gets changed immediately after the `DlsFlsLicenseRequestInterceptor` has been run, we might spam the logs with WARN messages and I'm not sure if it's worth it since the very next requests will be simply rejected.

Reply:
We synced and concluded that it is not necessary to log WARN messages.